### DISCUSSION TITLE
Vite Support is in Beta!

### DISCUSSION THREAD

**Discussion by jacksteamdev** [crxjs/chrome-extension-tools#136] [Announcements]
I'm excited to announce that RPCE now supports Vite! It's been quite a journey, and I'm grateful to everyone who helped get it ready for public consumption. :sweat_smile: 

### Get started in 90 seconds

Check out this how-to guide on DEV Community to get started: [Create a Vite-React Chrome Extension in 90 seconds](https://dev.to/jacksteamdev/create-a-vite-react-chrome-extension-in-90-seconds-3df7)

### Please contribute :handshake: 

Vite support will stay under the beta tag on NPM while we squash any remaining bugs. Please don't hesitate to [create an issue](https://github.com/extend-chrome/rollup-plugin-chrome-extension/issues/new/choose) if you run into something, and if you can, please contribute.

### New releases

Watch this discussion for news and release announcements! :rocket:

**Comment by onikienko** [comment]
That's great. Thank you for that monumental work. 

Unfortunately, it still doesn't work on Windows. Looks like I only one who use Windows here 🥲 
Maybe will try to debug and take a look later on. At least run tests and take a look at how it will be.

**Reply by jacksteamdev** [reply]
Sorry to hear that! Windows support is essential.

Can you open an issue when you have time so we can get it sorted out?

**Reply by onikienko** [reply]
Just opened #137

**Comment by jacksteamdev** [comment]
Version 4.0.1-11 just hit NPM!

This release adds support for CSS imports in content scripts!  Since Vite does the heavy lifting for CSS imports, we get PostCSS, CSS Modules, and CSS Pre-processor support in content scripts for free! In addition, the content script is using proper CSS, not CSS-in-JS, so your Chrome Extension will be faster.

**Reply by jacksteamdev** [reply]
I just discovered that if you're using Svelte, the CSS in your Svelte files gets the same treatment: Vite writes it to the assets folder, and RPCE adds it to the content script in the manifest. Amazing what you get for free!

**Comment by jacksteamdev** [comment]
Version 4.0.1-12 just hit NPM!

This release adds support for the Chrome Scripting API, Vite-style static-asset imports, and importing HTML files.

[See the release notes for more details](https://github.com/extend-chrome/rollup-plugin-chrome-extension/releases/tag/v4.0.1-12).

**Reply by filipw01** [reply]
Hey, first a lot of thanks for all you hard work on that ❤️  
Are html and script imports supported for manifest v2? I'm in a process of migrating extension to vite and this release stopped bundling resources from `web_accessible_resources` from my manifest so I started using the new method of importing scripts and html files which result in the following error. What's funny is that it fine bundling mp3 file, but can't handle js or html
```
[crx:post-hijacked-hooks] The "path" argument must be of type string. Received null
error during build:
TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received null
    at new NodeError (internal/errors.js:322:7)
    at validateString (internal/validators.js:124:11)
    at Object.resolve (path.js:1074:7)
    at /Users/filip/projects/chilipiper-frontend/node_modules/rollup/dist/shared/rollup.js:14015:78
    at Array.map (<anonymous>)
    at getCollapsedSourcemap (/Users/filip/projects/chilipiper-frontend/node_modules/rollup/dist/shared/rollup.js:14015:37)
    at /Users/filip/projects/chilipiper-frontend/node_modules/rollup/dist/shared/rollup.js:14024:24
    at Array.map (<anonymous>)
    at collapseSourcemaps (/Users/filip/projects/chilipiper-frontend/node_modules/rollup/dist/shared/rollup.js:14024:10)
    at Chunk.render (/Users/filip/projects/chilipiper-frontend/node_modules/rollup/dist/shared/rollup.js:14893:19)
```

**Reply by filipw01** [reply]
When trying to reproduce I noticed that this error pops up when you have declared `build.sourcemap` or `build.minify` in `vite.config.js` it fails with the above error. I hope this helps someone one day

**Reply by quisido** [reply]
@filipw01 It was helpful! This also happens when `compilerOptions.mapRoot` is set to an empty string.

**Comment by jacksteamdev** [comment]
Versions 4.0.1-11 and 4.0.1-12 introduced some regressions regarding building HTML files.

A fix is in the works, you can expect a new release in the next 2-3 days.

**Comment by filipw01** [comment]
I'm having trouble with dynamic imports in `content_scripts`, we have a shared library of components that we want to code split in other projects, but in the extension, we don't really care. 

Code like this in shared component `const SnackbarInner = lazy(() => import('./SnackbarInner'))` throws 

```
[crx:post-hijacked-hooks] Invalid value "iife" for option "output.format" - UMD and IIFE output formats are not supported for code-splitting builds.
error during build:
Error: Invalid value "iife" for option "output.format" - UMD and IIFE output formats are not supported for code-splitting builds.
    at error (/Users/filip/projects/chilipiper-frontend/node_modules/rollup/dist/shared/rollup.js:160:30)
    at validateOptionsForMultiChunkOutput (/Users/filip/projects/chilipiper-frontend/node_modules/rollup/dist/shared/rollup.js:16147:16)
    at Bundle.generate (/Users/filip/projects/chilipiper-frontend/node_modules/rollup/dist/shared/rollup.js:15981:17)
    at async handleGenerateWrite (/Users/filip/projects/chilipiper-frontend/node_modules/rollup/dist/shared/rollup.js:23618:23)
    at async Object.regenerateBundle (/Users/filip/projects/chilipiper-frontend/node_modules/rollup-plugin-chrome-extension/lib/index-cjs.js:1755:3)
    at async Promise.all (index 1)
    at async Object.generateBundle (/Users/filip/projects/chilipiper-frontend/node_modules/rollup-plugin-chrome-extension/lib/index-cjs.js:1697:30)
    at async Object.generateBundle (/Users/filip/projects/chilipiper-frontend/node_modules/rollup-plugin-chrome-extension/lib/index-cjs.js:9639:11)
    at async Bundle.generate (/Users/filip/projects/chilipiper-frontend/node_modules/rollup/dist/shared/rollup.js:15998:9)
    at async handleGenerateWrite (/Users/filip/projects/chilipiper-frontend/node_modules/rollup/dist/shared/rollup.js:23618:23)
```

Do you know if that is an unimplemented feature, bug or my usage/config might be wrong?

**Comment by jacksteamdev** [comment]
@filipw01 It's an unimplemented feature :sweat_smile:

The next beta version uses ESM for content scripts (see #200), and we're adding support for code-splitting/dynamic imports in dynamic scripts in #206.

**Reply by jacksteamdev** [reply]
@filipw01 Thanks for sharing your use cases!

To make sure I understand:

- Are you declaring content scripts in the manifest?
  - Is this where you're using dynamic imports?
- Are you importing scripts using the script query? e.g.`import script from './content?script`
- Are you importing HTML files?

**Reply by filipw01** [reply]
@jacksteamdev I have a content script that directly imports a library that uses dynamic imports. The extension I migrated to vite was written with MV2, so I guess it's time to migrate it to MV3 soon.

I also noticed that in 4.0.1-12, when importing HTML files, their JS resources are ignored. Is this feature added in the big rewrite, or is that intended? What I ended up doing is that I added vite config entries for necessary scripts and changed script path in HTML files to match

**Reply by jacksteamdev** [reply]
> a content script that directly imports a library that uses dynamic imports

That should be OK since RPCE bundles content scripts straight to disk. However, HMR for content scripts is different from extension pages; changes to content scripts will trigger a full extension reload and then a refresh for each page with a content script. True HMR for content scripts is coming, but it requires a custom HMR client that I'm working on next week. 

> I also noticed that in 4.0.1-12, when importing HTML files, their JS resources are ignored.

HTML bundling suffered a regression, but it's fixed in 4.0.1-16. In addition, I've added e2e tests that will ensure nothing like that happens again.

**Comment by jacksteamdev** [comment]
We've gotten so much great feedback on Vite support over the last few weeks! Thanks to everyone who participated for taking the time to test this library and post issues. All of you have been a great help.

# Beta version 4.0.1-16 has just hit NPM!

I've updated the article on DEV with the new API; [check it out for more info](https://dev.to/jacksteamdev/create-a-vite-react-chrome-extension-in-90-seconds-3df7).

Here are some of the highlights:

## Better HTML output :page_facing_up: 

The most wide-reaching problem involved invalid or missing HTML build output. I found a way to fix this by pushing extension pages into Vite's HTML pipeline. This works great but left behind lots of dead code in RPCE. So I decided to do a rewrite using what we've learned in our previous attempts. 

The result is a much leaner codebase that is more stable and extendable. A new testing strategy uses snapshot testing to track how changes impact the output files, and E2E tests make sure everything works in the browser, including HMR.

## :sparkles:  New API :sparkles: 
Since Vite supports importing JSON and TypeScript into config files, I've delegated manifest discovery and loading to Vite. Here's what it looks like now:

```javascript
// vite.config.js
import { crx } from 'rollup-plugin-chrome-extension'
import { defineConfig } from 'vite'
import manifest from './manifest.json'

export default defineConfig({
  plugins: [crx({ manifest })],
})
```

If you want to write your manifest in TypeScript, you can do it in the Vite config or import it from a separate file. RPCE exports a Vite-style define function now that provides types and IntelliSense.

```typescript
// manifest.config.ts
import { defineManifest } from 'rollup-plugin-chrome-extension'

export default defineManifest({
  // define your manifest
})
```

```typescript
// vite.config.ts
import { crx, defineManifest } from 'rollup-plugin-chrome-extension'
import { defineConfig } from 'vite'
import manifest from './manifest.config'

export default defineConfig({
  plugins: [crx({ manifest })],
})
```

Like Vite's `defineConfig`, `defineManifest` also can take a function that returns a manifest so you can extend the manifest based on runtime values.

```typescript
// manifest.config.ts
import { defineManifest } from 'rollup-plugin-chrome-extension'

export default defineManifest(({ command, mode }) => ({
  // manifest
}))
```

## :warning: Manifest Version 3 only :warning: 
Most people have requested MV3 support, and the Chrome Web Store is no longer accepting new MV2 extensions. As a result, this rewrite doesn't support Manifest Version 2.  

This will limit cross-browser support for now. However, all major browsers currently support or have announced support for MV3. 

#### State of MV3 cross-browser support as of 2022-02-14:
- Firefox has announced MV3 support in the coming months.
- Safari 15.4 beta already has MV3 support. 
- Edge supports everything Chrome supports.

After other more pressing features are supported, like true HMR for content scripts, we'll consider adding MV2 support.

## :cry: No more Rollup  
I've dropped Rollup support so we can ship this faster. I haven't seen any use cases for Rollup that Vite doesn't cover, and supporting both requires duplicating much of Vite's functionality. 

The plan is to change the NPM package name before releasing a stable version.

**Comment by jacksteamdev** [comment]
RPCE v4.0.1-18 just hit NPM! I'm totally pumped about this:

# True HMR in Content Scripts

The browser loads content scripts from the file system, so Vite's HMR server doesn't work. This release ports Vite's HMR to the file system. It took a lot of sweat to get all the pieces working, but it's here!

Instructions here: [Real Vite-React HMR in Chrome Extension Content Scripts - DEV Community 👩‍💻👨‍💻](https://dev.to/jacksteamdev/real-vite-react-hmr-in-chrome-extension-content-scripts-40di)

**Comment by jpozo20** [comment]
Thanks for all your hard work @jacksteamdev 👏🏻

Recently  I started to develop an extension to scratch my own itch and found out about this project by googling how to build one with Vite. Initially I couldn't get to make the imports work and only the code in the main file (main.tsx or index.tsx) was actually working, any other code which would be imported wasn't working and emitting an error about the [preamble detection](https://github.com/vitejs/vite/tree/main/packages/plugin-react#middleware-mode). After two days trying to make it work I discovered that this is solved by adding [this code](https://github.com/vitejs/vite/issues/1984#issuecomment-778289660) into a separate script since MV3 doesn't support inline scripts. 

For now, the dev build and HMR are working but there is an issue when running `yarn build` saying it can't resolve the entry module.
I tried many things adding options here and there but in the end nothing did work, and it looks like it's exclusive of RPCE cuz if you remove the crx from the plugins list then the `yarn build` builds fine.  I attach both the debug build and my vite.config.js. 

<details>
<summary>Build debug log</summary>

```
`vite:config bundled config file loaded in 504.49ms +0ms
  vite:config using resolved config: {
  vite:config   plugins: [
  vite:config     'alias',
  vite:config     'vite:react-babel',
  vite:config     'vite:react-refresh',
  vite:config     'vite:react-jsx',
  vite:config     'crx:content-scripts-pre',
  vite:config     'crx:dynamic-scripts',
  vite:config     'crx:manifest-init',
  vite:config     'crx:manifest-loader',
  vite:config     'crx:stub-input',
  vite:config     'vite:modulepreload-polyfill',
  vite:config     'vite:resolve',
  vite:config     'vite:html-inline-script-proxy',
  vite:config     'vite:css',
  vite:config     'vite:esbuild',
  vite:config     'vite:json',
  vite:config     'vite:wasm',
  vite:config     'vite:worker',
  vite:config     'vite:asset',
  vite:config     'vite:define',
  vite:config     'vite:css-post',
  vite:config     'vite:build-html',
  vite:config     'commonjs',
  vite:config     'vite:data-uri',
  vite:config     'rollup-plugin-dynamic-import-variables',
  vite:config     'vite:asset-import-meta-url',
  vite:config     'crx:hmr',
  vite:config     'crx:resources',
  vite:config     'crx:content-scripts-post',
  vite:config     'crx:dynamic-scripts',
  vite:config     'crx:background-loader-file',
  vite:config     'vite:build-import-analysis',
  vite:config     'vite:esbuild-transpile',
  vite:config     'vite:manifest',
  vite:config     'vite:reporter',
  vite:config     'vite:load-fallback',
  vite:config     'crx:manifest-post'
  vite:config   ],
  vite:config   build: {
  vite:config     target: [ 'es2019', 'edge88', 'firefox78', 'chrome87', 'safari13.1' ],
  vite:config     polyfillModulePreload: true,
  vite:config     outDir: 'dist',
  vite:config     assetsDir: 'assets',
  vite:config     assetsInlineLimit: 4096,
  vite:config     cssCodeSplit: true,
  vite:config     cssTarget: [ 'es2019', 'edge88', 'firefox78', 'chrome87', 'safari13.1' ],
  vite:config     sourcemap: 'inline',
  vite:config     rollupOptions: { input: [Object] },
  vite:config     minify: 'esbuild',
  vite:config     terserOptions: {},
  vite:config     write: true,
  vite:config     emptyOutDir: null,
  vite:config     manifest: true,
  vite:config     lib: false,
  vite:config     ssr: false,
  vite:config     ssrManifest: false,
  vite:config     reportCompressedSize: true,
  vite:config     chunkSizeWarningLimit: 500,
  vite:config     watch: null,
  vite:config     commonjsOptions: { include: [Array], extensions: [Array] },
  vite:config     dynamicImportVarsOptions: { warnOnError: true, exclude: [Array] }
  vite:config   },
  vite:config   resolve: {
  vite:config     dedupe: [ 'react', 'react-dom', 'react', 'react-dom' ],
  vite:config     alias: [ [Object], [Object] ]
  vite:config   },
  vite:config   optimizeDeps: {
  vite:config     include: [ 'react/jsx-dev-runtime', 'react/jsx-dev-runtime' ],
  vite:config     esbuildOptions: { keepNames: undefined, preserveSymlinks: undefined }
  vite:config   },
  vite:config   configFile: '/Users/jefry.pozo/Documents/DevEnv/projects/chrome/coma/vite.config.ts',
  vite:config   configFileDependencies: [ 'public/manifest.json', 'vite.config.ts' ],
  vite:config   inlineConfig: {
  vite:config     root: undefined,
  vite:config     base: undefined,
  vite:config     mode: undefined,
  vite:config     configFile: undefined,
  vite:config     logLevel: undefined,
  vite:config     clearScreen: undefined,
  vite:config     build: {}
  vite:config   },
  vite:config   root: '/Users/jefry.pozo/Documents/DevEnv/projects/chrome/coma',
  vite:config   base: '/',
  vite:config   publicDir: '/Users/jefry.pozo/Documents/DevEnv/projects/chrome/coma/public',
  vite:config   cacheDir: '/Users/jefry.pozo/Documents/DevEnv/projects/chrome/coma/node_modules/.vite',
  vite:config   command: 'build',
  vite:config   mode: 'production',
  vite:config   isProduction: true,
  vite:config   server: { fs: { strict: undefined, allow: [Array] } },
  vite:config   env: { BASE_URL: '/', MODE: 'production', DEV: false, PROD: true },
  vite:config   assetsInclude: [Function: assetsInclude],
  vite:config   logger: {
  vite:config     hasWarned: false,
  vite:config     info: [Function: info],
  vite:config     warn: [Function: warn],
  vite:config     warnOnce: [Function: warnOnce],
  vite:config     error: [Function: error],
  vite:config     clearScreen: [Function: clearScreen],
  vite:config     hasErrorLogged: [Function: hasErrorLogged]
  vite:config   },
  vite:config   createResolver: [Function: createResolver]
  vite:config } +7ms

vite v2.6.14 building for production...
✓ 0 modules transformed.
Could not resolve entry module (popup.html).
error during build:
Error: Could not resolve entry module (popup.html).
    at error (/Users/jefry.pozo/.npm/lib/node_modules/vite/node_modules/rollup/dist/shared/rollup.js:158:30)
    at ModuleLoader.loadEntryModule (/Users/jefry.pozo/.npm/lib/node_modules/vite/node_modules/rollup/dist/shared/rollup.js:22357:20)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Promise.all (index 0)`
```
</details>

<details>
<summary>Vite.config.js</summary>

```javascript
import { defineConfig } from "vite";
import { resolve } from "path";
import react from "@vitejs/plugin-react";

import { crx } from "rollup-plugin-chrome-extension";
import manifest from "./public/manifest.json";

// https://vitejs.dev/config/
export default defineConfig({
	plugins: [
		react(),
		crx({
			manifest,
			contentScripts: {
				preambleCode: false,
			},
		}),
	],
	build: {
		sourcemap: "inline",
		outDir: "dist",
		rollupOptions: {
			input: {
				main: resolve(__dirname, "popup.html"),
			},
		},
	},
});
```
</details>

**Reply by jpozo20** [reply]
I've been working on my project until the build is fixed since it doesn't affect the development time, but I found another peculiar bug when using tailwind.css. Again, I tested by using the normal workflow with Vite without RPCE and it works fine. 
For some reason, when using tailwind any file which uses a .css will have that CSS as a ModuleNode in its `importers` array and this was causing a stack overflow (or max call stack exceeded as it's called in javascript) because the CSS file in turn will have that file in its `importers` array.  

It's my first time debugging an NPM lib but the workaround is quite trivial. Because the latest releases (v4.0.1-18) are not available in the repo and previous versions weren't working (couldn't resolve imported scripts in HTML files), I'll attach the relevant code here, but you can tell me if you prefer a pull request.


In [`/src/node/plugin-hmr.ts`](https://github.com/extend-chrome/rollup-plugin-chrome-extension/blob/main/src/node/plugin-hmr.ts)
```javascript
function isImporter(file: string) {
    const seen = [];
    const pred = (node: ModuleNode) => {
	if (node.file === file) {
		return true;
	}
	seen.push(node);
	for (const node2 of node.importers) {
	    if (seen.indexOf(node) < 0) {
	        if (pred(node2)) {
		    return true;
		}
            }
        }
    };
return pred;
}

```

**Reply by jacksteamdev** [reply]
@jpozo20 Thanks for debugging this; `isImporter` looks good!

> Error: Could not resolve entry module (popup.html)

It's hard to give advice without more details. It could be something with the manifest. Could you open an issue? 🙏

**Reply by jpozo20** [reply]
After doing some more test I found the problem with tailwind.css only occurs when using a background script, otherwise it works fine.  Though I think the fix should be applied just in case. 

> It's hard to give advice without more details. It could be something with the manifest. Could you open an issue? 🙏

After looking through the issues I saw your comment about paths being relative to the viteconfig.root in the manifest. It built fine after doing that 👍.
Still It would be good if you uploaded the v4.0.1-18 for future debugging since previous versions have other issues with HTML.

**Reply by jpozo20** [reply]
Unfortunately, there are still some issues with CSS when using tailwind.css, I opened the issue #274

**Reply by jpozo20** [reply]
@jacksteamdev is there a way to get access to the latest beta source code? I need to be applying the fix above every time I install or upgrade a package since npm reverses the original file. With the source at least I could link RPCE to the I have modified on disk so my fix would stay between installs.

**Reply by jacksteamdev** [reply]
The `main` branch is up to date now. If you clone `main` and link it should work.

**Comment by Toumash** [comment]
Out of beta. We're stable now! 🎉 https://github.com/crxjs/chrome-extension-tools/releases/tag/%40crxjs%2Fvite-plugin%402.0.0

## [🎉 Vite support is officially out of beta! 🎉](https://github.com/crxjs/chrome-extension-tools/discussions/1020)
